File: perturbation_assignment/unified_analysis.R
#' Check if Docker is Available and Return Its Path
#'
#' The `has_docker` function checks if the Docker executable is available
#' in the system's `PATH`. It returns a list with a boolean indicating if Docker
#' is found, and the path to the Docker executable if it is available.
#'
#' @return
#' A list with two elements:
#' \item{found}{A logical value: `TRUE` if Docker is available in the system's
#' `PATH`, `FALSE` otherwise.}
#' \item{path}{A character string containing the full path to the Docker
#' executable if found,  or an empty string if not found.}
#'
#' @examples
#' result <- has_docker()
#' if (result$found) {
#'   cat("Docker is available at:", result$path, "\n")
#' } else {
#'   cat("Docker is not available.\n")
#' }
#'
#' @export
has_docker <- function() {
  path <- Sys.which("docker")
  return(list(found = nzchar(path), path = path, fn = run_in_docker))
}
#' Check if the script is running in Docker.
#'
#' @returns A truthy value indicating the state.
#' @export
is_running_in_docker <- function() {
  dockerenv_exists <- file.exists("/.dockerenv")
  cgroup_exists <- file.exists("/proc/1/cgroup")
  in_container_runtime <- FALSE
  if (cgroup_exists) {
    in_container_runtime <- any(
      grepl("docker", readLines("/proc/1/cgroup", warn = FALSE))
    )
  }
  return(dockerenv_exists || in_container_runtime)
}
#' Run a docker container.
#'
#' @param image_name The docker image you want to run.
#' @param volumes The list of volumes to mount to the container.
#' @param additional_arguments Vector of arguments to pass to the container.
#'
#' @export
run_in_docker <- function(image_name,
                          volumes = list(),
                          additional_arguments = c()) {
  base_command <- "run --privileged=true --platform linux/amd64 --rm"
  for (volume in volumes) {
    volume[1] <- normalizepath::normalize_path(volume[1],
      path_mappers = c(normalizepath::docker_mount_mapper)
    )
    base_command <- paste(base_command, "-v", paste(
      volume[1],
      volume[2],
      sep = ":"
    ))
  }
  base_command <- paste(base_command, image_name)
  for (argument in additional_arguments) {
    base_command <- paste(base_command, argument)
  }
  system2("docker", args = base_command, stdout = "", stderr = "")
}

#' unified_analysis
#'
#' @description Unified CIRI workflow for Single and Dual guide analysis
#' @param scripts_dir Host directory containing the R scripts
#' @param data_dir Host directory containing the input data
#' @param mode Analysis Mode: 1=Single Guide, 2=Dual Guide (allowed values: 1, 2)
#' @return Results of the operation
#'
#' @export
unified_analysis <- function(scripts_dir,
data_dir,
mode) {
  # Type validation
  if (!is.character(scripts_dir) || length(scripts_dir) != 1) {
    stop("scripts_dir must be a single character string")
  }
  if (!is.character(data_dir) || length(data_dir) != 1) {
    stop("data_dir must be a single character string")
  }
  valid_mode <- c("1", "2")
  if (!is.character(mode) || length(mode) != 1 || !(mode %in% valid_mode)) {
    stop(paste0("mode must be one of: ", paste(valid_mode, collapse=", ")))
  }
  
  # Security checks
  if (grepl("\\.\\./|\\.\\\\|\\/\\.\\./|\\\\\\.\\\\\\.\\\\", scripts_dir)) {
    stop("Path traversal detected in scripts_dir")
  }
  if (grepl("\\.\\./|\\.\\\\|\\/\\.\\./|\\\\\\.\\\\\\.\\\\", data_dir)) {
    stop("Path traversal detected in data_dir")
  }
  
  # Check if directory exists
  if (!is_running_in_docker()) {
    if (!dir.exists(scripts_dir)) {
      stop(paste("scripts_dir:", scripts_dir, "does not exist"))
    }
  }
  
  # Check if directory exists
  if (!is_running_in_docker()) {
    if (!dir.exists(data_dir)) {
      stop(paste("data_dir:", data_dir, "does not exist"))
    }
  }
  
  # Process file paths for Docker volume mounting
  # Process scripts_dir for Docker
  scripts_dir_abspath <- normalizePath(scripts_dir, mustWork = FALSE)
  scripts_dir_dir <- dirname(scripts_dir_abspath)
  scripts_dir_filename <- basename(scripts_dir)
  # Process data_dir for Docker
  data_dir_abspath <- normalizePath(data_dir, mustWork = FALSE)
  data_dir_dir <- dirname(data_dir_abspath)
  data_dir_filename <- basename(data_dir)
  
  # Main volume mount point
  main_mount_dir <- scripts_dir_dir
  
  # Execute Docker container with error handling
  tryCatch({
    result <- run_in_docker(
      image_name = "docker.io/hedgelab/ciri_analysis",
      volumes = list(
        c(scripts_dir_dir, "/scripts"),
        c(data_dir_dir, "/data")
      ),
      additional_arguments = c(
        "Rscript",
        "/scripts/CIRI_unified.R",
        "/data",
        mode,
      )
    )
    
    # Process result
    return(list(
      status = "success",
      output_dir = file.path(main_mount_dir, "unified_analysis_results")
    ))
  }, error = function(e) {
    stop(paste("Docker execution failed:", e$message))
  })
}

File: perturbation_assignment/unified_analysis.py
#!/usr/bin/env python3

import os
import sys
import re
import subprocess
import pathlib
import logging
from typing import Dict, List, Any, Optional, Union
from dataclasses import dataclass

# Configure logging
logger = logging.getLogger(__name__)

@dataclass
class Result:
  status: str
  output_dir: str
  message: str = ""

def validate_path(path: str) -> str:
  """Validate and normalize a file path."""
  if not path:
    raise ValueError("Path cannot be empty")
  return os.path.abspath(os.path.expanduser(path))

def is_running_in_docker() -> bool:
  """Check if we're running inside a Docker container."""
  return os.path.exists('/.dockerenv')

def run_docker(image: str, volumes: Dict[str, str], env: Dict[str, str], args: List[str]) -> str:
  """Run a Docker container with specified parameters."""
  cmd = ['docker', 'run', '--rm']
  
  for src, dst in volumes.items():
    cmd.extend(['-v', f"{src}:{dst}"])
  
  for key, val in env.items():
    cmd.extend(['-e', f"{key}={val}"])
  
  cmd.append(image)
  cmd.extend(args)
  
  logger.info(f"Running Docker command: {' '.join(cmd)}")
  result = subprocess.run(cmd, capture_output=True, text=True, check=False)
  
  if result.returncode != 0:
    logger.error(f"Docker execution failed: {result.stderr}")
    raise RuntimeError(f"Docker execution failed: {result.stderr}")
  
  return result.stdout

def unified_analysis(scripts_dir: str, data_dir: str, mode: str) -> Result:
  """
  Unified CIRI workflow for Single and Dual guide analysis
  
  Parameters:
      scripts_dir: Host directory containing the R scripts
      data_dir: Host directory containing the input data
      mode: Analysis Mode: 1=Single Guide, 2=Dual Guide (allowed values: 1, 2)
  
  Returns:
      Result: Results of the operation
  """
  # Parameter validation
  if not isinstance(scripts_dir, str):
    raise TypeError(f"scripts_dir must be a string, got {type(scripts_dir).__name__}")
  scripts_dir_path = validate_path(scripts_dir)
  if not isinstance(data_dir, str):
    raise TypeError(f"data_dir must be a string, got {type(data_dir).__name__}")
  data_dir_path = validate_path(data_dir)
  mode_valid_values = ["1", "2"]
  if not isinstance(mode, str):
    raise TypeError(f"mode must be a string, got {type(mode).__name__}")
  if mode not in mode_valid_values:
    raise ValueError(f"mode must be one of {mode_valid_values}")
  
  # Directory existence checks
  if not is_running_in_docker():
    if not os.path.isdir(scripts_dir_path):
      raise NotADirectoryError(f"Directory {scripts_dir_path} does not exist")
  if not is_running_in_docker():
    if not os.path.isdir(data_dir_path):
      raise NotADirectoryError(f"Directory {data_dir_path} does not exist")
  
  # Process file paths for Docker volume mounting
  # Process scripts_dir for Docker
  scripts_dir_abspath = os.path.abspath(scripts_dir_path if 'scripts_dir_path' in locals() else scripts_dir)
  scripts_dir_dir = os.path.dirname(scripts_dir_abspath)
  scripts_dir_filename = os.path.basename(scripts_dir)
  # Process data_dir for Docker
  data_dir_abspath = os.path.abspath(data_dir_path if 'data_dir_path' in locals() else data_dir)
  data_dir_dir = os.path.dirname(data_dir_abspath)
  data_dir_filename = os.path.basename(data_dir)
  
  # Main volume mount point
  main_mount_dir = scripts_dir_dir
  
  # Execute Docker container with error handling
  try:
    # Prepare Docker volumes
    volumes = {}
    volumes[scripts_dir_dir] = "/scripts"
    volumes[data_dir_dir] = "/data"
    
    # Prepare environment variables
    env_vars = {}
    
    # Prepare Docker arguments
    docker_args = []
    docker_args.append("Rscript")
    docker_args.append("/scripts/CIRI_unified.R")
    docker_args.append("/data")
    docker_args.append(str(mode))
    
    # Run Docker container
    run_docker("docker.io/hedgelab/ciri_analysis", volumes, env_vars, docker_args)
    
    # Create results directory
    output_dir = os.path.join(main_mount_dir, "unified_analysis_results")
    os.makedirs(output_dir, exist_ok=True)
    
    return Result(status="success", output_dir=output_dir)
  except Exception as e:
    logger.error(f"Docker execution failed: {str(e)}")
    return Result(status="error", output_dir="", message=str(e))


if __name__ == "__main__":
  import argparse
  
  parser = argparse.ArgumentParser(description="Unified CIRI workflow for Single and Dual guide analysis")
  parser.add_argument('--scripts_dir', help="Host directory containing the R scripts")
  parser.add_argument('--data_dir', help="Host directory containing the input data")
  parser.add_argument('--mode', choices=["1", "2"], help="Analysis Mode: 1=Single Guide, 2=Dual Guide")
  
  args = parser.parse_args()
  
  result = unified_analysis(
    scripts_dir=args.scripts_dir,
    data_dir=args.data_dir,
    mode=args.mode,
  )
  
  print(f"Status: {result.status}")
  if result.status == "success":
    print(f"Output directory: {result.output_dir}")
  else:
    print(f"Error: {result.message}")
    sys.exit(1)
